addons/payment_stripe_sca/models/payment.py

```diff
@@ -0,0 +1,275 @@
+            res = res.get("charges").get("data")[0]
+
+        _logger.info("_stripe_create_payment_intent: Values received:\n%s", pprint.pformat(res))
+        return res
+
+    def stripe_s2s_do_transaction(self, **kwargs):
+        self.ensure_one()
+        result = self._stripe_create_payment_intent(acquirer_ref=self.payment_token_id.acquirer_ref, email=self.partner_email)
+        return self._stripe_s2s_validate_tree(result)
+
+    def _create_stripe_refund(self):
+        refund_params = {
+            "charge": self.acquirer_reference,
+            "amount": int(float_round(self.amount * 100, 2)),  # by default, stripe refund the full amount (we don't really need to specify the value)
+            "metadata[reference]": self.reference,
+        }
+
+        _logger.info("_create_stripe_refund: Sending values to stripe URL, values:\n%s", pprint.pformat(refund_params))
+        res = self.acquirer_id._stripe_request("refunds", refund_params)
+        _logger.info("_create_stripe_refund: Values received:\n%s", pprint.pformat(res))
+        return res
+
+    @api.model
+    def _stripe_form_get_tx_from_data(self, data):
+        """ Given a data dict coming from stripe, verify it and find the related
+        transaction record. """
+        reference = data.get("reference")
+        if not reference:
+            stripe_error = data.get("error", {}).get("message", "")
+            _logger.error(
+                "Stripe: invalid reply received from stripe API, looks like "
+                "the transaction failed. (error: %s)",
+                stripe_error or "n/a",
+            )
+            error_msg = _("We're sorry to report that the transaction has failed.")
+            if stripe_error:
+                error_msg += " " + (_("Stripe gave us the following info about the problem: '%s'") % stripe_error)
+            error_msg += " " + _(
+                "Perhaps the problem can be solved by double-checking your "
+                "credit card details, or contacting your bank?"
+            )
+            raise ValidationError(error_msg)
+
+        tx = self.search([("reference", "=", reference)])
+        if not tx:
+            error_msg = _("Stripe: no order found for reference %s") % reference
+            _logger.error(error_msg)
+            raise ValidationError(error_msg)
+        elif len(tx) > 1:
+            error_msg = _("Stripe: %s orders found for reference %s") % (len(tx), reference)
+            _logger.error(error_msg)
+            raise ValidationError(error_msg)
+        return tx[0]
+
+    def _stripe_s2s_validate_tree(self, tree):
+        self.ensure_one()
+        if self.state != "draft":
+            _logger.info("Stripe: trying to validate an already validated tx (ref %s)", self.reference)
+            return True
+
+        status = tree.get("status")
+        tx_id = tree.get("id")
+        vals = {"date_validate": fields.datetime.now(), "acquirer_reference": tx_id}
```

Comment:
does this work? shouldn't it be `fields.Datetime.now()`?

Reply:
Done